### PR TITLE
[WIP] use withCredentials XMLHTTPRequest appropriately

### PR DESCRIPTION
**Pull request by camallen**
Use the underlying credentialed / non-credentialed http requests in https://github.com/zooniverse/json-api-client/pull/52 to avoid CORS issues for open resources on the API.

The underlying issue is that an API change to cors implementation in https://github.com/zooniverse/Panoptes/pull/3199 ensures that all public (wildcard) origins disallows credentialed requests, i.e. our API allows `{ origins: '*', resource: 'api/*'}` so these resources can't be accessed with credentialed requests. As it stands our `json-api-client` always attaches uses `withCredentials` which is a risk of leaking private data.

### CODE REVIEW

File: lib/csrf-token.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+function getCSRFToken() {
+  console.log('Getting csrf token');
+  var url = config.oauthHost + '/users/sign_in/?now=' + Date.now();
```
Comment (eatyourgreens) [inline lib/csrf-token.js]: This used `config.host` in the original code but I'm not sure if that was correct. I assume this change is to allow for a different auth host from the Panoptes API host?

Comment (camallen) [reply lib/csrf-token.js]: possibly - that was me copy/pasting. I assumed the config was shared, static data

I assume this file should include `config` lib as well then, might be good to refactor `JSON_HEADERS` into config as well for re-use.

Comment (srallen) [inline lib/csrf-token.js]: The host has changed here.

Comment (eatyourgreens) [reply lib/csrf-token.js]: Oh yeah, looks like this needs to import the config too. I missed that.

Comment (eatyourgreens) [reply lib/csrf-token.js]: Config is static, by the way. It's just a JS dictionary exported from `config.js`.

Comment (srallen) [reply lib/csrf-token.js]: It might not be an issue. Should confirm in manual testing.

Comment (srallen) [reply lib/csrf-token.js]: I'll make the changes and push a commit up

File: lib/auth.js [hunk 2]
```diff
@@ -139,7 +126,7 @@ module.exports = new Model({
         }.bind(this));
       } else {
         console.log('Registering new account', given.login);
-        var registrationRequest = this._getAuthToken().then(function(token) {
+        var registrationRequest = getCSRFToken(url, JSON_HEADERS, apiClient).then(function(token) {
```
Comment (srallen) [inline lib/auth.js]: The `getCSRFToken` function doesn't take params... was this an intended refactor or in error?

Comment (srallen) [reply lib/auth.js]: We can just require in the `apiClient` and the config file after the `JSON_HEADERS` move to it. The `url` is undefined here, but the `getCSRFToken` function defines it within.

File: package.json [hunk 3]
```diff
@@ -6,7 +6,7 @@
   "author": "Zooniverse",
   "license": "Apache-2.0",
   "dependencies": {
-    "json-api-client": "^4.0.2",
+    "json-api-client": "file:../json-api-client",
```
Comment (srallen) [inline package.json]: I'm going to revert this change. I had trouble getting the right json-api-client loaded up with PFE when testing locally. Using npm link with both PJC and this dep got the right version of everything loaded for manual testing for me.

Comment (srallen) [reply package.json]: Actually no, the combination of this and then npm linking them both to PFE is what was working. This will need to be reverted for merge and publishing though.

File: test/auth.js [hunk 4]
```diff
@@ -16,7 +16,7 @@ test('Checking the current user initially fails', function(t) {
     });
 });
 
-test('Registering an account with no data fails', function(t) {
+test.only('Registering an account with no data fails', function(t) {
```
Comment (srallen) [inline test/auth.js]: This will need to be removed prior to merge.

File: lib/csrf-token.js (hunk 1 above)
Comment (camallen) [reply lib/csrf-token.js]: I know understand the concern re the change of host, my apologies. I'll look into this, it should respect the different host config options.

File: package.json (hunk 3 above)
Comment (camallen) [reply package.json]: I think we'll need a release candidate of https://github.com/zooniverse/json-api-client/pull/52 to test this properly. What's the best of of dealing this that?

**Comment by camallen** [comment]
I have no faith in these tests assocaited with the repo, they share state between test runs and tests must be run in order for them to work (if any fail above others below will fail). Ideally we fix these as well (i thought they'd work) but i can't figure out what's going on with them.

**Comment by srallen** [comment]
Since we do not trust the tests, I suggest we remove them.

**Comment by srallen** [comment]
We can remove the tests in a separate PR to keep this scoped.

File: package.json (hunk 3 above)
Comment (srallen) [reply package.json]: I've published json-api-client@5.0.0-rc.0 and have updated the dependency here.
